iquip/apps/builder.py

```diff
@@ -0,0 +1,419 @@
+        """Extended.
+        
+        Args:
+            default: The default value. If it does not exist, it is set to False.
+        """
+        super().__init__(name, parent=parent)
+        # widgets
+        self.checkBox = QCheckBox(self)
+        self.checkBox.setChecked(default)
+        # layout
+        self.layout.addWidget(self.checkBox)
+
+    def value(self) -> bool:
+        """Overridden.
+        
+        Returns the status of the checkBox.
+        """
+        return self.checkBox.isChecked()
+
+
+class _EnumerationEntry(_BaseEntry):
+    """Entry class for an enumeration value."""
+
+    def __init__(
+        self,
+        name: str,
+        choices: List[str],
+        default: Optional[str] = None,
+        parent: Optional[QWidget] = None
+    ):
+        """Extended.
+        
+        Args:
+            choices: The pre-defined candidates.
+            default: The default value. If it does not exist, it is set to the first candidate.
+        """
+        super().__init__(name, parent=parent)
+        # widgets
+        self.comboBox = QComboBox(self)
+        for choice in choices:
+            self.comboBox.addItem(choice)
+        if choices:
+            self.comboBox.setCurrentText(choices[0] if default is None else default)
+        # layout
+        self.layout.addWidget(self.comboBox)
+
+    def value(self) -> str:
+        """Overridden.
+        
+        Returns the value of the comboBox.
+        """
+        return self.comboBox.currentText()
+
+
+class _NumberEntry(_BaseEntry):
+    """Entry class for a number value.
+    
+    Attributes:
+        scale: The scale factor that actually applies.
+        type: The type of the value. If "int", value() returns an integer value.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        unit: str,
+        scale: float,
+        step: float,
+        min: Optional[float],  # pylint: disable=redefined-builtin
+        max: Optional[float],  # pylint: disable=redefined-builtin
+        ndecimals: int,
+        type: str,  # pylint: disable=redefined-builtin
+        default: Optional[float] = None,
+        parent: Optional[QWidget] = None
+    ):  # pylint: disable=too-many-arguments
+        """Extended.
+        
+        Args:
+            unit: The unit of the value.
+            step: The step between values changed by the up and down button.
+            min: The minimum value. None for no minimum.
+            max: The maximum value. None for no maximum.
+            ndecimals: The maximum number of decimals.
+            default: The default value. If it does not exist, it is set to the min value.
+            scale, type: See the attributes section in _NumberEntry.
+        """
+        super().__init__(name, parent=parent)
+        self.scale = scale
+        self.type = type
+        # widgets
+        self.spinBox = QDoubleSpinBox(self)
```

Comment:
Are you going to use `QDoubleSpinBox` regardless of the type (int, float)?
Moreover, even if the type is int, is it okay to give non-zero `decimals`?

Reply:
> Are you going to use QDoubleSpinBox regardless of the type (int, float)?

Yes. I think it isn't a problem to use `QDoubleSpinBox` when the type is int.

Reply:
> Moreover, even if the type is int, is it okay to give non-zero decimals?

In that case, the proxy server would return an error becuase it isn't allowed in ARTIQ.

Reply:
So if I set a number value with `type="int"` and `ndecimals=1` in an experiment class, what happens when I fetch the arguments of that experiment class?

Reply:
In that case, the proxy server causes the exception as below.

<img width="80%" src="https://github.com/snu-quiqcl/iquip/assets/76851886/5ce94d02-f0ee-4a3d-84ff-5cb920691d4f">

On the server side, the log is as follows.

![image](https://github.com/snu-quiqcl/iquip/assets/76851886/1281774c-3311-4b2e-b610-83d57501b290)

Reply:
Thank you for the screenshots! That confirms that ARTIQ rejects such ill-formatted experiment codes.